packages/aws-cdk-lib/core/lib/stage.ts

```diff
@@ -210,16 +216,62 @@ export class Stage extends Construct {
    * calls will return the same assembly.
    */
   public synth(options: StageSynthesisOptions = { }): cxapi.CloudAssembly {
-    if (!this.assembly || options.force) {
+
+    let newConstructPaths = this.listAllConstructPaths(this);
+
+    // If the assembly cache is uninitiazed, run synthesize and reset construct paths cache
+    if (this.constructPathsCache.size == 0 || !this.assembly || options.force) {
       this.assembly = synthesize(this, {
         skipValidation: options.skipValidation,
         validateOnSynthesis: options.validateOnSynthesis,
       });
+      newConstructPaths = this.listAllConstructPaths(this);
+      this.constructPathsCache = newConstructPaths;
     }
 
+    // If the construct paths set has changed
+    if (!this.constructPathSetsAreEqual(this.constructPathsCache, newConstructPaths)) {
+      const errorMessage = 'The construct tree has been modified after synthesis. Only the results of the first synth() call are written to disk, and modifications done after it are ignored. Avoid construct tree mutations after synth() has been called.';
+      if (options.errorOnDuplicateSynth ?? true) {
+        throw new Error(errorMessage);
```

Comment:
Not sure we can really use the exact same message. the call to action is different when there is an error vs a warning. In my head, the error scenario is this:

- Hey we have detected that you called synth() multiple times and changed the construct tree after the first synth()
- You cannot do this, we are failing the execution.
- If you really know what you are doing, you can set this flag to keep going: <enter flag here>